petab/v2/lint.py

```diff
@@ -0,0 +1,549 @@
+    """A task to validate the condition table of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.condition_df is None:
+            return
+
+        try:
+            check_condition_df(
+                problem.condition_df,
+                model=problem.model,
+                observable_df=problem.observable_df,
+                mapping_df=problem.mapping_df,
+            )
+        except AssertionError as e:
+            return ValidationError(str(e))
+
+
+class CheckObservableTable(ValidationTask):
+    """A task to validate the observable table of a PEtab problem."""
+
+    def run(self, problem: Problem):
+        if problem.observable_df is None:
+            return
+
+        try:
+            check_observable_df(
+                problem.observable_df,
+            )
+        except AssertionError as e:
+            return ValidationIssue(
+                level=ValidationIssueSeverity.ERROR, message=str(e)
+            )
+
+
+class CheckObservablesDoNotShadowModelEntities(ValidationTask):
+    """A task to check that observable IDs do not shadow model entities."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.observable_df is None or problem.model is None:
+            return
+
+        shadowed_entities = [
+            obs_id
+            for obs_id in problem.observable_df.index
+            if problem.model.has_entity_with_id(obs_id)
+        ]
+        if shadowed_entities:
+            return ValidationError(
+                f"Observable IDs {shadowed_entities} shadow model entities."
+            )
+
+
+class CheckParameterTable(ValidationTask):
+    """A task to validate the parameter table of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.parameter_df is None:
+            return
+
+        try:
+            df = problem.parameter_df
+            _check_df(df, PARAMETER_DF_REQUIRED_COLS[1:], "parameter")
+
+            if df.index.name != PARAMETER_ID:
+                return ValidationError(
+                    f"Parameter table has wrong index {df.index.name}."
+                    f" Expected {PARAMETER_ID}.",
+                )
+
+            check_ids(df.index.values, kind="parameter")
+
+            for column_name in PARAMETER_DF_REQUIRED_COLS[
+                1:
+            ]:  # 0 is PARAMETER_ID
+                if not np.issubdtype(df[column_name].dtype, np.number):
+                    assert_no_leading_trailing_whitespace(
+                        df[column_name].values, column_name
+                    )
+
+            # nominal value is generally optional, but required if any for any
+            #  parameter estimate != 1
+            non_estimated_par_ids = list(
+                df.index[
+                    (df[ESTIMATE] != 1)
+                    | (
+                        pd.api.types.is_string_dtype(df[ESTIMATE])
```

Comment:
:hankey:

Ideally this would be a validation error

Reply:
That's the old problem that data types are undefined. I guess we could change that in v2, but that's a separate discussion. We also have the issue, for example, in the conditions table. Depending on the contents, the columns will be object or float.

Reply:
https://github.com/PEtab-dev/libpetab-python/issues/290